centreon/src/Core/Notification/Infrastructure/Repository/DbNotifiableResourceFactory.php

```diff
@@ -0,0 +1,211 @@
+namespace Core\Notification\Infrastructure\Repository;
+
+use Core\Notification\Application\Converter\{NotificationHostEventConverter, NotificationServiceEventConverter};
+use Core\Notification\Domain\Model\{NotifiableHost, NotifiableResource, NotifiableService, NotificationHostEvent};
+
+class DbNotifiableResourceFactory
+{
+    /**
+     * @param array<int,array{
+     *  notification_id: int,
+     *  host_id: int,
+     *  host_name: string,
+     *  host_alias: string|null,
+     *  host_events: int,
+     *  service_id: int,
+     *  service_name: string,
+     *  service_alias: string,
+     *  service_events: int,
+     *  included_service_events: int
+     * }> $records
+     *
+     * @throws \Throwable
+     *
+     * @return \Generator|null
+     */
+    public static function createFromRecords(iterable $records): \Generator
+    {
+        $currentNotificationId = 0;
+        $currentRecords = [];
+        foreach ($records as $record) {
+            if ($currentNotificationId === 0) {
+                $currentNotificationId = $record['notification_id'];
+                $currentRecords[] = $record;
+                continue;
+            }
+
+            if ($currentNotificationId === $record['notification_id']) {
+                $currentRecords[] = $record;
+                continue;
+            }
+
+            yield self::createNotifiableResourceFromRecord($currentNotificationId, $currentRecords);
+
+            $currentRecords = [];
+            $currentRecords[] = $record;
+            $currentNotificationId = $record['notification_id'];
+        }
+
+        if ([] !== $currentRecords) {
+            yield self::createNotifiableResourceFromRecord($currentNotificationId, $currentRecords);
+        }
+
+    }
+
+    /**
+     * @param int $notificationId
+     * @param array<int,array{
+     *  notification_id: int,
+     *  host_id: int,
+     *  host_name: string,
+     *  host_alias: string|null,
+     *  host_events: int,
+     *  service_id: int,
+     *  service_name: string,
+     *  service_alias: string,
+     *  service_events: int,
+     *  included_service_events: int
+     * }> $records
+     *
+     * @throws \Throwable
+     *
+     * @return NotifiableResource
+     */
+    private static function createNotifiableResourceFromRecord(int $notificationId, array $records): NotifiableResource
+    {
+        $notificationHosts = [];
+        $currentHostId = 0;
+        $currentRecords = [];
+        $currentHostEvents = [];
+        $index = 0;
+        foreach ($records as $record) {
+            if ($currentHostId === 0) {
+                $currentHostId = $record['host_id'];
+                $currentRecords[] = $record;
+                $index++;
+                continue;
+            }
+
+            if ($currentHostId === $record['host_id']) {
+                $currentRecords[] = $record;
+                $index++;
+                continue;
+            }
+
+            if ($record['host_events'] !== '0') {
```

Comment:
Please could you replace the '0' by a constant to make code clearer

Reply:
Done